### DISCUSSION TITLE
Unable to acquire lock on session with ID

### DISCUSSION THREAD

**Discussion by Zruto** [nextflow-io/nextflow#3987] [Q&A]
When I do not apply -resume, I can submit multiple nextflow tasks at the same time, and can generate multiple session IDs in the cache of .nextflow. But when I use -resume, even if only one task is submitted, it will appear "Unable to acquire lock on session with ID" error.What is the problem?
![image](https://github.com/nextflow-io/nextflow/assets/46238825/ce03d9f9-14f5-4ff4-b00e-15d6d27d10de)

**Comment by bentsherman** [comment]
In general it is not possible to run multiple instances of Nextflow from the same launch directory, because each instance will acquire a lock on the `.nextflow` folder. But also, if a workflow run fails in a really weird way, it might leave behind the lock file it created, which prevents subsequent runs. In that case you can just delete the lock file from the `.nextflow` folder.

**Reply by Zruto** [reply]
What I mean is that I submit nextflow one by one, that is, when the previous nextflow is running, I submit another task. But when using -resume, there is a problem of Unable to acquire lock on session with ID

**Reply by Zruto** [reply]
Specifically, once I create nextflow with - resume, the generated session ID is always the same

`2023-06-01 19:58:25	7.3s	mad_poitras	ERR	edd9642e328011ab8688d48d0c957cf7	fe2126e8-4b28-4b5f-8c8e-4fb377052316`
`2023-06-01 20:02:12	9.8s	berserk_liskov	ERR	8e900b2d1fa6966a5770b111c757ddee	fe2126e8-4b28-4b5f-8c8e-4fb377052316`
`2023-06-01 20:05:09	8.3s	determined_lalande	ERR	25489f9e13b93fac57df605572dbd9a1	fe2126e8-4b28-4b5f-8c8e-4fb377052316`

**Reply by Zruto** [reply]
Through testing, it was found that after using - resume, the generated session ID is the same as the one you last generated?
`2023-06-01 20:19:15	7.5s	jolly_murdock	ERR	cbcff53b78a0526886e5186e8e2f35c5	fe2126e8-4b28-4b5f-8c8e-4fb377052316	nextflow *** -resume`
`2023-06-01 20:19:54	7.4s	sharp_khorana	ERR	cbcff53b78a0526886e5186e8e2f35c5	f94966e8-9483-4a97-a9d9-686d2ef18162	nextflow ***`
`2023-06-01 20:24:58	7.3s	lethal_volta	ERR	ced097532dd2e0a8b6dd717e3142411c	f94966e8-9483-4a97-a9d9-686d2ef18162	nextflow *** -resume`

**Reply by bentsherman** [reply]
Yes, Nextflow will try to use the same session ID on a resume. But that doesn't mean that you can launch multiple instances from the same directory. If you want to run many workflows concurrently, you need to launch each workflow from a different directory. That way each instance will have it's own `.nextflow` cache.

**Reply by Zruto** [reply]
Isn't the calculation method of sessionID calculated and generated based on the information in the input file? Even if I changed the content of the input file, when using resume, the same sessionID as the last command was generated. Why does it use the same sessionID as the last command?

**Reply by bentsherman** [reply]
The session ID is just a unique random id. But on a resume it will try to use a previous session id.

**Reply by bentsherman** [reply]
Interesting tidbit I learned today, Nextflow maintains a separate cache db for each session ID. So that may be why you can have multiple concurrent instances in the same directory as long as you don't resume. Although I think you would still hit a lock issue with the history file, which there is only one for all sessions.

**Reply by Oksanak22** [reply]
@bentsherman by saying "If you want to run many workflows concurrently, you need to launch each workflow from a different directory" do you mean I have to have a nextflow installed in a different folder?

**Reply by bentsherman** [reply]
no, you just need to launch each pipeline in a different directory

**Reply by Oksanak22** [reply]
@bentsherman Thank you. I will run the pipeline with different datasets and drop the output in different folders.

**Reply by Oksanak22** [reply]
Hello @bentsherman.

I submitted 2 datasets and dropped the outputs into different folders. (I am running the analysis on a Beluga cluster)
Unfortunately, it didn't work.

N E X T F L O W  ~  version 23.04.3
Launching `main.nf` [zen_heyrovsky] DSL2 - revision: a648196f77
ERROR ~ Unable to acquire lock on session with ID e34d851f-d75a-4d7a-aa9f-91607e644b88

Common reasons for this error are:
 - You are trying to resume the execution of an already running pipeline
 - A previous execution was abruptly interrupted, leaving the session open

You can see which process is holding the lock file by using the following command:
 - lsof /lustre04/scratch/wf-pore-c-0.1.0/wf-pore-c/.nextflow/cache/e34d851f-d75a-4d7a-aa9f-91607e644b88/db/LOCK

 -- Check '.nextflow.log' file for details

**Reply by bentsherman** [reply]
Did you launch them from different directories?

**Comment by Oksanak22** [comment]
Hello, @bentsherman!

Could you, please explain to me how to launch from a different directory if
I have wf-pore-c-0.1.0 installed just in one place and you told me I don't
need to install a WF in other folder.

Best,
Oksana

On Tue, 12 Dec 2023 at 10:25, Ben Sherman ***@***.***> wrote:

> Did you launch them from different directories?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextflow-io/nextflow/discussions/3987#discussioncomment-7832163>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXZT5EVPGWNKZSLX7OB6METYJBZOXAVCNFSM6AAAAAAYWHMWU6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQMZSGE3DG>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Oksana

**Reply by bentsherman** [reply]
Oh I see what you mean now. The best practice here is to not clone the pipeline code manually. Instead you can do something like:
```bash
nextflow run epi2me-labs/wf-pore-c -r 0.1.0
```

in any directory, and Nextflow will clone the repo into `$HOME/.nextflow/assets` for you. This way you could launch many pipelines from many different directories:
```bash
cd run1 ; nextflow run epi2me-labs/wf-pore-c -r 0.1.0 --input ./inputs1
cd run2 ; nextflow run epi2me-labs/wf-pore-c -r 0.1.0 --input ./inputs2
# ...
```

Each run will use the same copy of your pipeline code in `~/.nextflow` but will otherwise be independent of each other

**Reply by Oksanak22** [reply]
Thank you @bentsherman for the explanations. 
For my previous run, I used this script and submitted as a bash file on a cluster at this location
/home/scratch/wf-pore-c-0.1.0/wf-pore-c

as follow:

module load apptainer/1.1.8 java/17.0.2 nextflow/23.04.3
nextflow run main.nf -c nextflow.config \

 --fastq /home/scratch/45G.fastq \
 --ref /home/scratch/no_unitigs_genome.fasta \
 --cutter HindIII \
 --mcool_resolutions 1000,2000,3000 \
 --pairs  --chromunity --mcool \
--out_dir 45G \
 -profile apptainer

**If I understand you correctly now I need to modify it like this for the r1:**

module load apptainer/1.1.8 java/17.0.2 nextflow/23.04.3


 nextflow run /home/oksanak/scratch/wf-pore-c-0.1.0/wf-pore-c/main.nf -r 0.1.0 -c nextflow.config \

 --fastq /home/scratch/45G.fastq \
 --ref /home/scratch/no_unitigs_genome.fasta \
 --cutter HindIII \
 --mcool_resolutions 1000,2000,3000 \
 --pairs  --chromunity --mcool \
--out_dir 45G \
 -profile apptainer

**And for the r2 i have to be somewhere other than /home/scratch/wf-pore-c-0.1.0/wf-pore-c and submit the following .sh script:**

module load apptainer/1.1.8 java/17.0.2 nextflow/23.04.3

nextflow run /home/oksanak/scratch/wf-pore-c-0.1.0/wf-pore-c/main.nf -r 0.1.0 -c nextflow.config \

 --fastq /home/scratch/60G.fastq \
 --ref /home/scratch/no_unitigs_genome.fasta \
 --cutter HindIII \
 --mcool_resolutions 1000,2000,3000 \
 --pairs  --chromunity --mcool \
--out_dir 60G \
 -profile apptainer

Am I right?

**Reply by bentsherman** [reply]
sure that should work

**Reply by Oksanak22** [reply]
Hello @bentsherman 

Unfortunately, it didn't work.
I got the same error for the both runs:

 N E X T F L O W  ~  version 23.04.3
Revision option cannot be used when running a local script

**Reply by bentsherman** [reply]
Because you have already downloaded the pipeline code and are using the absolute path of the main script, you don't need to specify `-r 0.1.0`. Sorry I didn't catch that before

You should either do:
```bash
# it will download the repo for you into $HOME/.nextflow/assets
nextflow run epi2me-labs/wf-pore-c -r 0.1.0
```

or
```bash
nextflow run /home/oksanak/scratch/wf-pore-c-0.1.0/wf-pore-c/main.nf
```

**Comment by Oksanak22** [comment]
Hi, @bentsherman!

I was able to start runs, but now for both of them I have an error:

[e3/86e1b2] NOTE: Process `POREC:fastcat (1)` terminated with an error exit status (1) -- Error is ignored

executor >  slurm (1)
[e3/86e1b2] process > POREC:fastcat (1)              [100%] 1 of 1, failed: 1 ✔
[-        ] process > POREC:index_ubam               -
[65/dede14] process > POREC:prepare_genome:index_... [100%] 1 of 1, cached: 1 ✔
[ae/2f3916] process > POREC:prepare_genome:index_... [100%] 1 of 1, cached: 1 ✔
[-        ] process > POREC:digest_align_annotate    -
[-        ] process > POREC:haplotag_alignments      -
[-        ] process > POREC:merge_coordsorted_bams   -
[-        ] process > POREC:merge_namesorted_bams    -
[-        ] process > POREC:create_restriction_bed   -
[-        ] process > POREC:to_pairs_file            -
[-        ] process > POREC:pairsToCooler            -
[-        ] process > POREC:merge_mcools             -
[-        ] process > POREC:merge_pairs              -
[-        ] process > POREC:merge_pairs_stats        -
[-        ] process > POREC:pair_stats_report        -
[-        ] process > POREC:merge_parquets_to_dat... -
[f9/660e35] process > POREC:getVersions              [100%] 1 of 1, cached: 1 ✔
[7d/5f7d3e] process > POREC:getParams                [100%] 1 of 1, cached: 1 ✔
[-        ] process > POREC:makeReport               -
[-        ] process > POREC:prepare_hic              -

Completed at: 14-Dec-2023 16:20:59
Duration    : 3m 24s
CPU hours   : 0.1 (52.9% cached, 47.1% failed)
Succeeded   : 0
Cached      : 4
Ignored     : 1
Failed      : 1

Even I resume run, i keep getting this error.

**Reply by bentsherman** [reply]
It does not give you a more specific error message?

**Reply by Oksanak22** [reply]
No, nothing more

**Reply by bentsherman** [reply]
I guess it's because your error strategy is set to "ignore" so it does not print the error message

**Reply by Oksanak22** [reply]
Hello, @bentsherman . Could you, please, help me either change the error strategy to print the error or how I can see the error?
Thank you

**Reply by bentsherman** [reply]
You can set the error strategy in your nextflow config:
```groovy
process.errorStrategy = 'terminate'
```

**Reply by Oksanak22** [reply]
Thank you @bentsherman . 
I changed config file as you told me, and it failed again.
Even in .nextflow.log 

Dec-18 12:47:51.850 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260382; id: 1755; name: POREC:pairsToCooler (577); status: COMPLETED; exit>Dec-18 12:47:51.902 [main] DEBUG nextflow.trace.WorkflowStatsObserver - Workflow completed > WorkflowStats[succeededCount=548; failedCount=4; ignoredCount=0; cachedCount=1182; pendingCou>Dec-18 12:47:51.902 [main] DEBUG nextflow.trace.TraceFileObserver - Workflow completed -- saving trace file
Dec-18 12:47:51.904 [main] DEBUG nextflow.trace.ReportObserver - Workflow completed -- rendering execution report
Dec-18 12:47:52.046 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260383; id: 1756; name: POREC:pairsToCooler (578); status: COMPLETED; exit>Dec-18 12:47:52.059 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260384; id: 1757; name: POREC:to_pairs_file (582); status: COMPLETED; exit>Dec-18 12:47:52.067 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260385; id: 1758; name: POREC:to_pairs_file (583); status: COMPLETED; exit>Dec-18 12:47:52.075 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260386; id: 101; name: POREC:pairsToCooler (15); status: COMPLETED; exit: >Dec-18 12:47:52.094 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260387; id: 110; name: POREC:pairsToCooler (17); status: COMPLETED; exit: >Dec-18 12:47:52.101 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260388; id: 106; name: POREC:pairsToCooler (16); status: COMPLETED; exit: >Dec-18 12:47:52.118 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260389; id: 1608; name: POREC:pairsToCooler (464); status: COMPLETED; exit>Dec-18 12:47:52.124 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260390; id: 1609; name: POREC:pairsToCooler (465); status: COMPLETED; exit>Dec-18 12:47:52.145 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260391; id: 1610; name: POREC:pairsToCooler (466); status: COMPLETED; exit>Dec-18 12:47:52.236 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260392; id: 1611; name: POREC:pairsToCooler (467); status: COMPLETED; exit>Dec-18 12:47:52.443 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 43260393; id: 1612; name: POREC:pairsToCooler (468); status: COMPLETED; exit>Dec-18 12:47:52.807 [Task submitter] DEBUG nextflow.executor.GridTaskHandler - [SLURM] submitted process POREC:pairsToCooler (574) > jobId: 43260394; workDir: /lustre04/scratch/oksanak/w>Dec-18 12:47:52.808 [Task submitter] INFO  nextflow.Session - [a6/ddbf1b] Submitted process > POREC:pairsToCooler (574)
Dec-18 12:47:53.798 [main] DEBUG nextflow.trace.ReportObserver - Execution report summary data:
  [{"cpuUsage":{"mean":2.33,"min":2.33,"q1":2.33,"q2":2.33,"q3":2.33,"max":2.33,"minLabel":"POREC:fastcat (1)","maxLabel":"POREC:fastcat (1)","q1Label":"POREC:fastcat (1)","q2Label":"POR>Dec-18 12:47:55.475 [main] DEBUG nextflow.trace.TimelineObserver - Workflow completed -- rendering execution timeline
Dec-18 12:47:57.784 [main] DEBUG nextflow.cache.CacheDB - Closing CacheDB done
Dec-18 12:47:58.386 [main] DEBUG nextflow.script.ScriptRunner - > Execution complete -- Goodbye
 

But in  slurm.out:

ERROR ~ Error executing process > 'POREC:makeReport (1)'

Caused by:
  Process `POREC:makeReport (1)` terminated with an error exit status (127)

**Reply by bentsherman** [reply]
yes, the goal of changing the error strategy was to make it show a more informative error message. Is there not more detail in the `slurm.out` ?

**Comment by Oksanak22** [comment]
This is the tail of slurm.out


Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details
WARN: Killing running tasks (14)

executor >  slurm (566)
[53/2645a6] process > POREC:fastcat (1)              [100%] 1 of 1, cached: 1 ✔
[48/c7cf6e] process > POREC:index_ubam (1)           [100%] 1 of 1, cached: 1 ✔
[4c/4629f5] process > POREC:prepare_genome:index_... [100%] 1 of 1, cached: 1 ✔
[a6/1d450c] process > POREC:prepare_genome:index_... [100%] 1 of 1, cached: 1 ✔
[69/d5d8ec] process > POREC:digest_align_annotate... [100%] 584 of 584, cache...
[-        ] process > POREC:haplotag_alignments      -
[-        ] process > POREC:merge_coordsorted_bams   -
[-        ] process > POREC:merge_namesorted_bams    -
[f5/0da924] process > POREC:create_restriction_be... [100%] 1 of 1, cached: 1 ✔
[5d/36ecdf] process > POREC:to_pairs_file (583)      [ 99%] 578 of 582, cache...
[12/d82af0] process > POREC:pairsToCooler (468)      [ 98%] 561 of 568, cache...
[-        ] process > POREC:merge_mcools             -
[-        ] process > POREC:merge_pairs              -
[-        ] process > POREC:merge_pairs_stats        -
[-        ] process > POREC:pair_stats_report        -
[-        ] process > POREC:merge_parquets_to_dat... -
[de/d1c3dc] process > POREC:getVersions              [100%] 1 of 1, cached: 1 ✔
[35/e2ddda] process > POREC:getParams                [100%] 1 of 1, cached: 1 ✔
[f9/163116] process > POREC:makeReport (1)           [100%] 4 of 4, failed: 4...
[-        ] process > POREC:prepare_hic              -
ERROR ~ Error executing process > 'POREC:makeReport (1)'

Caused by:
  Process `POREC:makeReport (1)` terminated with an error exit status (127)

Command executed:

  echo '[
      {
          "barcode": null,
          "type": "test_sample",
          "alias": "C1-II"
      }
  ]' > metadata.json
      workflow-glue report wf-pore-c-report.html         --metadata metadata.json         --stats per_read_stats.tsv   >
Command exit status:
  127

Command output:
  (empty)

Command error:
  INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
  .command.sh: line 9: workflow-glue: command not found

Work dir:
  /lustre04/scratch/wf-pore-c/C1-II/work/f9/163116d696d1790a4a5648e5323dab

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details
[a6/ddbf1b] Submitted process > POREC:pairsToCooler (574)

**Reply by bentsherman** [reply]
There's your error:
```
.command.sh: line 9: workflow-glue: command not found
```

**Reply by Oksanak22** [reply]
I see this error now. But I can't understand why I got this error if I just changed a dataset. And I don't see a place where I can change the work directory and even in .command.sh I see the same info:

#!/bin/bash -euo pipefail
echo '[
    {
        "barcode": null,
        "type": "test_sample",
        "alias": "C107-II"
    }
]' > metadata.json
    workflow-glue report wf-pore-c-report.html         --metadata metadata.json         --stats per_read_stats.tsv         --versions versions         --params params.json
